record_generator.go

```diff
@@ -0,0 +1,119 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package generator
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math/rand"
+	"os"
+	"time"
+
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/google/uuid"
+)
+
+type recordGenerator interface {
+	generate() (sdk.Record, error)
+}
+
+type defaultRecordGen struct {
+	config RecordConfig
+}
+
+func (g defaultRecordGen) generate() (sdk.Record, error) {
+	p, err := g.generatePayload(g.config)
+	if err != nil {
+		return sdk.Record{}, err
+	}
+	return sdk.Record{
+		Position:  []byte(uuid.New().String()),
+		Metadata:  make(map[string]string),
+		Key:       sdk.RawData(uuid.NewString()),
+		Payload:   p,
+		CreatedAt: time.Now(),
+	}, nil
+}
+
+func newRecordGenerator(config RecordConfig) defaultRecordGen {
+	return defaultRecordGen{config: config}
+}
+
+func (g defaultRecordGen) generatePayload(config RecordConfig) (sdk.Data, error) {
+	switch config.FormatType {
+	case FormatFile:
+		return g.generateFilePayload(config.FormatOptions["path"].(string))
+	case FormatRaw, FormatStructured:
+		return g.generateStruct(config.FormatType, config.FormatOptions["fields"].(map[string]string))
+	default:
+		return nil, fmt.Errorf("unrecognized type of payload to generate: %q", config.FormatType)
+	}
+}
+
+func (g defaultRecordGen) generateFilePayload(path string) (sdk.Data, error) {
+	bytes, err := os.ReadFile(path)
```

Comment:
That's true. I expect most use cases to be either:
1. Many small records
2. Not so many large records

We can also have a few small ones (but that is not really interesting), or a many large ones (that's probably realistic). Small records can be easily generated, and it's maybe even easier to use them that way then to load them from files. Given that, I've focused on not so many large payloads, where disk reads aren't that important, but memory is. I'm fine with making caching configurable though.

Reply:
Is there a downside if the connector always cached the file?

Reply:
Well on the other hand it could lower the frequency of generating records. A quick benchmark shows that reading a 50mb file takes ~0.1s on average on my machine. Depending on the size of the file in the performance benchmarks this could be a bottleneck.